### PR TITLE
merge when green

### PR DESCRIPTION
**Pull request by hjdivad**
- [ ] test in `nyx-example` using a workflow like `merge-when-green.yml`
- [ ] merge: check checks as well as statuses
- [ ] report failure by context
- [ ] finish todo tests
  - commit in multiple branches
  - commit from fork (need to get owner from status.branches; maybe parsing the URL)
  - commit in multiple prs (probably just try to merge both; it might actually work since one might be downstream of the other)
- [ ] cleanup

---

- [x] readContext
- [x] merge(ctx)
- [x] run -> merge(ctx)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,111 +1,7 @@
 #!/usr/bin/env node
 
-import * as yargs from 'yargs';
-import setupHardRejection from 'hard-rejection';
-import reportFailure from './commands/report-failure';
-import merge from './commands/merge';
-import assert from './utils/assert';
+// This file is mainly for indirection w/the shebang so it's valid JavaScript
+// The real implementation is in ./run.ts
+import main from './run';
```
Comment (hjdivad) [inline src/index.ts:5]: This is slightly annoying but I don't know how to get the typescript langserver to be happy with a typescript file that has a shebang

File: package.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
     "build:ts": "tsc",
     "build:ts:defs": "tsc --declaration --emitDeclarationOnly",
     "build:watch": "tsc --watch",
+    "repl": "ts-node -D 7017 -D 2393",
```
Comment (hjdivad) [inline package.json:19]: This actually works pretty nice with `neoterm` and the right `.vimrc`.  Note that there's a couple of issues:

- [ts-node missing a recovery code](https://github.com/TypeStrong/ts-node/pull/1014)
- `ts-node` will still fail in annoying ways for some multiline expressions.  You basically want to turn off format on save for `.repl.ts`

Here's my current project hackety hack vimrc as it relates to this
```vim
" exploring repl workflows
" TODO: don't autoformat the repl file
"   requires adding 1007 to ts-node recovery codes
"     https://github.com/TypeStrong/ts-node/pull/1014
"   but this is still not enough due to `gh.pulls.list({` giving the wrong
"   error
" 
" Right now depends on manually doing
"   :2Tnew
"   " check the id of the new terminal
"   :Tls
"   :TREPLSetTerm id_of_repl_term
" 
function s:EditREPL()
  " setup REPL
  "   if REPL setup
  "     ensure visible
  "   else
  "   - C-W-t (go top left)
  "   - C-W n (split)
  "   - get bufnr
  "   - get term id from bufnr
  "   - TREPLSetTerm id
  "   - open repl file
  "
  exe "edit .repl.ts"
endfunction

nmap <leader>re :call <SID>EditREPL()<CR>
vmap <leader>re :TREPLSendSelection<CR>

augroup REPL
  au!
  au BufWritePre .repl.ts exe 'TREPLSendFile'
augroup END

```

Comment (hjdivad) [inline package.json:19]: - `2393` is the important one (ignore duplicate function)
- sure would be nice to have comments in json eh?

File: src/run.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        assert(
+          !!token,
+          `nyx merge expects either the \`--token\` argument or for $GITHUB_AUTH to be set`
+        );
+
+        if (context === undefined) {
+          assert(
+            !!prInput && !!owner && !!repo,
+            `nyx merge expects either the \`--owner\`, \`--repo\`, and \`--pr\` arguments, or for $GITHUB_CONTEXT to be for a status or pull request review`
+          );
```
Comment (hjdivad) [inline src/run.ts:106]: @rwjblue can you think of a better way than to make the args optional (ie no `demandArgument`) and assert like this?

File: src/utils/read-context.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import assert from './assert';
+
+// TODO: rather than define these types partially we might want to read them from octokit similar to what is done in src/commands/merge
+// we'll want the single types (eg review vs reviews)
```
Comment (hjdivad) [inline src/utils/read-context.ts:4]: I looked *very briefly* at this but the types don't appear the same.  I don't know that Octokit actually has a type anywhere for the status events, so we may just have to build up these mini types.

File: src/utils/read-context.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    if (e instanceof SyntaxError && /JSON/.test(e.message)) {
+      throw new Error(
+        'process.env.GITHUB_CONTEXT found, but it is not a valid JSON-encoded string'
+      );
+    }
+
+    throw e;
+  }
+
+  // type guard would be better here if there's a reasonable one
```
Comment (hjdivad) [inline src/utils/read-context.ts:91]: ah well we have type guards now so we should probably throw here if the event is unexpected (ie not status and not pr review)

thoughts @rwjblue ?

File: src/run.ts (hunk 3 above)
Comment (rwjblue) [reply src/run.ts:106]: No, I think this is fine.

File: src/index.ts (hunk 1 above)
Comment (rwjblue) [reply src/index.ts:5]: Seems fine (though somewhat annoying).

File: src/run.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
+import * as yargs from 'yargs';
+import setupHardRejection from 'hard-rejection';
+import reportFailure from './commands/report-failure';
+import merge, { mergeByContext } from './commands/merge';
+import assert from './utils/assert';
+import readContext from './utils/read-context';
+
+export default function main(): void {
```
Comment (rwjblue) [inline src/run.ts:8]: Should probably async?

File: src/index.ts (hunk 1 above)
Comment (hjdivad) [reply src/index.ts:5]: @rwjblue if you can figure out how to do it i'm very happy to avoid the indirection

File: src/commands/merge.ts [hunk 7]
```diff
@@ -114,3 +122,62 @@ export default async function merge({
...
+  _mergeFn = merge,
+}: ContextArgs): Promise<ExitCode> {
+  let { owner, repo } = readRepository(context);
+  const github = new Octokit({
+    auth: token,
+    userAgent: '@malleatus/nyx automerger (by context)',
+  });
+
+  let event = context.event;
+  if (isStatusEvent(event)) {
```
Comment (rwjblue) [inline src/commands/merge.ts:146]: I'm curious why you went with this specific approach. Specifically, why not have a `getPullRequestNumberFromContext` kinda function (moving this logic into that function and leaving this file focused on the logic of doing the merging...

File: src/utils/read-context.ts (hunk 5 above)
Comment (rwjblue) [reply src/utils/read-context.ts:91]: Ya, seems good. We could use `never` + a conditional to have TS help us enforce (e.g. if we add another type to the union for `GitHubContext`).

Comment (hjdivad) [reply src/utils/read-context.ts:91]: @rwjblue i'm not following. How would we use `never` with a conditional here?  Can you illustrate in a code snippet?

File: src/commands/merge.ts (hunk 7 above)
Comment (hjdivad) [reply src/commands/merge.ts:146]: That seems like a perfectly good refactor to me

So we'd end up with something like this?
```
export async function mergeByContext({
  context,
  token,

  _mergeFn = merge
}) {
  let prNumber = getPullRequestNumberFromContext(context);
  if(prNumber === undefined) {
    return ExitCode.DidNotRun;
  }

  return _mergeFn({owner, repo, prNumber, token });
}
```

**Comment by hjdivad** [comment]
We probably want a debug flag with logging too

**Comment by hjdivad** [comment]
@rwjblue okay the issue is that the github actions workflows don't use the status API: they use the checks API.  Our tests were passing because we *were* using the status api.

So we'll want to update the merge logic to compare `status ∪ checks`

**Pull request merged**